# Урок 2. Метрики качества классификации ч.1

В уроке мы разбирали матрицу ошибок и некоторые метрики, основанные на терминах матрицы ошибок. Здесь же посмотрим, как эти метрики применять на практике и что  по ним можно выяснить.

### 1.1
Выкачайте тренировочный датасет Titanic - train.csv - с сайта [kaggle](https://www.kaggle.com/c/titanic/data). С помощью функции pd.read_csv() загрузите данные в датафрейм. Выведите первые 20 строк и проанализируйте данные: какие колонки присутствуют (более конкретная информация по ним есть на сайте kaggle), каким образом в них обозначены данные и какие типы данных используются (используйте pandas.dtypes). 

In [122]:
import pandas as pd

In [123]:
df_titanic = pd.read_csv('titanic\\train.csv')
df_titanic.head(20)

PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   
12           13         0       3   
13           14         0       3   
14           15         0       3   
15           16         1       2   
16           17         0       3   
17           18         1       2   
18           19         0       3   
19           20         1       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
5                                    Moran, Mr. James    male   NaN      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
12                     Saundercock, Mr. William Henry    male  20.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
17                       Williams, Mr. Charles Eugene    male   NaN      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                            Masselmani, Mrs. Fatima  female   NaN      0   

    Parch            Ticket     Fare Cabin Embarked  
0       0         A/5 21171   7.2500   NaN        S  
1       0          PC 17599  71.2833   C85        C  
2       0  STON/O2. 3101282   7.9250   NaN        S  
3       0            113803  53.1000  C123        S  
4       0            373450   8.0500   NaN        S  
5       0            330877   8.4583   NaN        Q  
6       0             17463  51.8625   E46        S  
7       1            349909  21.0750   NaN        S  
8       2            347742  11.1333   NaN        S  
9       0            237736  30.0708   NaN        C  
10      1           PP 9549  16.7000    G6        S  
11      0            113783  26.5500  C103        S  
12      0         A/5. 2151   8.0500   NaN        S  
13      5            347082  31.2750   NaN        S  
14      0            350406   7.8542   NaN        S  
15      0            248706  16.0000   NaN        S  
16      1            382652  29.1250   NaN        Q  
17      0            244373  13.0000   NaN        S  
18      0            345763  18.0000   NaN        S  
19      0              2649   7.2250   NaN        C

In [124]:
df_titanic.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

Train.csv будет содержать информацию о подгруппе пассажиров на борту (точнее, 891 человек) и, что важно, покажет, выжили они или нет, что также известно как «наземная правда».

Набор данных `test.csv` содержит аналогичную информацию, но не раскрывает «основную правду» для каждого пассажира. Это ваша работа, чтобы предсказать эти результаты.

Используя шаблоны, которые вы найдете в данных train.csv, предскажите, выжили ли остальные 418 пассажиров на борту (найденные в test.csv).

### 1.2
Проверьте, имеются ли пропущенные значения в колонках, и выведите сумму всех пропущенных значений в каждой из колонок.

In [125]:
df_titanic.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Мы видим что пропущенные значения в столбце Age 177 штук и в столбце Cabin - 687 штук, а так же в Embarked - 2 шт.

### 1.3
Замените все пропущенные значения колонки Age на медианы в зависимости от пола человека: т.е. если пол человека в строке с пропущенным значением "male", заменяете пропущенное значение возраста на медиану по всем известным возрастам мужчин, и наоборот. Выведите медианы возраста в зависимости от пола. Пропущенные значения колонок Cabin и Embarked замените на U (Unknown).

Примечание: для группировки по полу можно использовать метод df.groupby(), который имеет встроенные агрегатные функции (в т.ч. для вычисления медианы).  Для установки пропущенных значений в датасете воспользуйтесь функцией df.apply().

In [126]:
df_titanic_ = df_titanic.groupby('Sex').agg('median')['Age']
df_titanic_female_age = df_titanic_[0]
df_titanic_male_age = df_titanic_[1]
print(f'Медиана возраста у мужчин: {df_titanic_male_age}; \nМедиана возраста у женщин: {df_titanic_female_age}')

Медиана возраста у мужчин: 29.0; 
Медиана возраста у женщин: 27.0


In [127]:
values = {"Cabin": 'U', "Embarked": 'U'}
df_titanic = df_titanic.fillna(value=values)

In [128]:
df_titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500     U        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250     U        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500     U        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000     U        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500     U        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500     U        Q  

[891 rows x 12 columns]

Создадим функцию которая заполняет пропущенные значения у мужчин и женщим медианным возрвстом

In [130]:
def function(row):
    if None == row.Age:
        if 'male' == row.Sex:
            return df_titanic_male_age
        else:
            return df_titanic_female_age

df_titanic['Age'] = df_titanic.apply(function, axis = 1)
df_titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  29.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  27.0      1   
2                               Heikkinen, Miss. Laina  female  27.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  27.0      1   
4                             Allen, Mr. William Henry    male  29.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  29.0      0   
887                       Graham, Miss. Margaret Edith  female  27.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  27.0      1   
889                              Behr, Mr. Karl Howell    male  29.0      0   
890                                Dooley, Mr. Patrick    male  29.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500     U        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250     U        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500     U        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000     U        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500     U        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500     U        Q  

[891 rows x 12 columns]

### 1.4
Выведите возраст пассажиров с PassengerID = [6, 20]. Убедитесь, что заполнены все пропущенные значения (воспользуйтесь функцией df.isnull() ).

In [131]:
data = df_titanic[(df_titanic['PassengerId'] >= 6) & (df_titanic['PassengerId'] <= 20)]['Age']
data

5     29.0
6     29.0
7     29.0
8     27.0
9     27.0
10    27.0
11    27.0
12    29.0
13    29.0
14    27.0
15    27.0
16    29.0
17    29.0
18    27.0
19    27.0
Name: Age, dtype: float64

In [132]:
data.isnull().sum()

0

Так как не все классификаторы могут работать с категориальными признаками, в данном случае необходимо заменить их на числовые.

Также можно было бы выделить расширенные признаки, такие как статус пассажира (Mr., Mrs., Miss., Dr., Master. и т.д.), собственный индекс билета, индекс палубы и другие, но здесь ограничимся выбором более простых признаков.

### 1.5
В колонке Sex замените значения на 0, если пол "male", и на 1, если "female". В колонке Embarked замените параметры "U", "S", "C", "Q" на 0, 1, 2, 3 соответственно. Отбросьте колонки PassengerId, Name, Ticket, Cabin. Выведите первые 20 строк получившегося набора данных.

In [133]:
def function(row):
    if row.Sex == 'male':
        return 0
    else:
        return 1

df_titanic['Sex'] = df_titanic.apply(function, axis = 1)
df_titanic.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    0  29.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1  27.0      1      0   
2                             Heikkinen, Miss. Laina    1  27.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500     U        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250     U        S

In [134]:
def function(row):
    if row.Embarked == 'U':
        return 0
    elif row.Embarked == 'S':
        return 1
    elif row.Embarked == 'C':
        return 2
    elif row.Embarked == 'Q':
        return 3

df_titanic['Embarked'] = df_titanic.apply(function, axis = 1)
df_titanic.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    0  29.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1  27.0      1      0   
2                             Heikkinen, Miss. Laina    1  27.0      0      0   

             Ticket     Fare Cabin  Embarked  
0         A/5 21171   7.2500     U         1  
1          PC 17599  71.2833   C85         2  
2  STON/O2. 3101282   7.9250     U         1

Отбросьте колонки PassengerId, Name, Ticket, Cabin. Выведите первые 20 строк получившегося набора данных.

In [135]:
df_titanic = df_titanic.drop(['PassengerId', 'Name', 'Ticket', 'Cabin'], axis = 1)
df_titanic

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0           0       3    0  29.0      1      0   7.2500         1
1           1       1    1  27.0      1      0  71.2833         2
2           1       3    1  27.0      0      0   7.9250         1
3           1       1    1  27.0      1      0  53.1000         1
4           0       3    0  29.0      0      0   8.0500         1
..        ...     ...  ...   ...    ...    ...      ...       ...
886         0       2    0  29.0      0      0  13.0000         1
887         1       1    1  27.0      0      0  30.0000         1
888         0       3    1  27.0      1      2  23.4500         1
889         1       1    0  29.0      0      0  30.0000         2
890         0       3    0  29.0      0      0   7.7500         3

[891 rows x 8 columns]

In [136]:
df_titanic.isnull().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64

### 1.6
Приступим к построению классификаторов. **Условимся, что, если функция или объект модели имеют параметр random_state, то устанавливаем его равным 17 в каждом из случаев.**

Разделите данные на тренировочный и тестовый датасеты, установив размер тестового как 0.25 (первая колонка Survived является целевой, поэтому необходимо сначала ее отделить от признаков).

In [137]:
from sklearn.model_selection import train_test_split 

In [138]:
y = df_titanic['Survived']
X = df_titanic.drop(['Survived'], axis = 1)

In [139]:
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size = 0.25 , random_state = 17)

print(f'X_train shape: {x_train.shape}, y_train shape: {y_train.shape},\n'
      f'X_test shape: {x_test.shape}, y_test shape: {y_test.shape}')

X_train shape: (668, 7), y_train shape: (668,),
X_test shape: (223, 7), y_test shape: (223,)


### 1.7
Обучите на полученных выборках несколько классификаторов, которые импортированы ниже.

**Примечание: в методе ближайших соседей используйте количество соседей, равное 5.**

In [140]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression

In [141]:
KKN = KNeighborsClassifier(n_neighbors = 5)
GNB = GaussianNB()
DTC = DecisionTreeClassifier()
LR = LogisticRegression()

In [142]:
kkn_model = KKN.fit(x_train, y_train)
gnb_model = GNB.fit(x_train, y_train)
dtc_model = DTC.fit(x_train, y_train)
lr_model = LR.fit(x_train, y_train)

C:\Users\pc\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\neighbors\base.py:908: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  self._y = np.empty(y.shape, dtype=np.int)
C:\Users\pc\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\tree\tree.py:163: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If yo

Очевидно, в этом случае использовать метрику accuracy, которую мы применяли до этого, не совсем правильно: она не даст верную оценку классификатору, потому как выживших явно меньше, чем погибших.

Вспомним про такие метрики как точность и полнота. Здесь они подойдут хорошо, так как у нас имеются положительный и отрицательный классы, и определить корректно один из них может оказаться более важным .

### 1.8
В каждом из пунктов следующего теста выберите, что важнее максимизировать: точность (precision) или полноту (recall). Выпишите ответы.

1. Вероятность того, что при определенной поломке самолета он сможет долететь до пункта назначения (1 - долетел, 0 - не долетел).
2. Предсказание, представляет ли человек опасность, по анализу психического состояния (1 - представляет опасность, 0 - не представляет опасности).
3. Предсказание ухода клиента (1 - клиент ушел, 0 - остался).
4. Выявление рака на основе медицинских показателей (1 - болен раком, 0 - здоров).
5. Предсказание летальности при наблюдаемой мутации (1 - выживание, 0 - летальный исход).
6. Определение важности происшествия для экстренных служб (1 - важно, 0 - неважно).
7. Окупятся ли вложения в бизнес (1 - окупятся, 0 - не окупятся).

Ваш ответ:

1. <br> Полнота
2. <br> Полнота
3. <br> Точность
4. <br> Полнота
5. <br> Полнота
6. <br> Полнота
7. <br> Точность

### 1.9
Определите, что важнее использовать в случае датасета Titanic: полноту или точность. Объясните, почему.


Ответ: 

Точность системы в пределах выживших – это доля людей действительно принадлежащих классу выживших относительно всех людей которых система отнесла к этому классу. 

Полнота системы – это доля найденных классфикатором людей принадлежащих классу выживших относительно всех людей этого класса в тестовой выборке.

Наверно нас интересует полнота. 

### 1.10
Для каждого классификатора выведите матрицу ошибок и самостоятельно рассчитайте метрику recall (расчеты должны присутствовать). Проверьте расчеты, воспользовавшись встроенной метрикой recall_score.

In [144]:
from sklearn.metrics import confusion_matrix, recall_score

In [146]:
kkn_model_pred = kkn_model.predict(x_test)
gnb_model_pred = gnb_model.predict(x_test)
dtc_model_pred = dtc_model.predict(x_test)
lr_model_pred = lr_model.predict(x_test)

C:\Users\pc\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\base.py:291: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  indices = (scores > 0).astype(np.int)


In [153]:
for x in [kkn_model_pred, gnb_model_pred, dtc_model_pred, lr_model_pred]:
    print(confusion_matrix(y_test, x))
    print(recall_score(y_test, x, average='macro'))
    print('\n')
    

[[115  16]
 [ 32  60]]
0.7650182542316628


[[114  17]
 [ 29  63]]
0.77750580816462


[[113  18]
 [ 32  60]]
0.7573846664454033


[[115  16]
 [ 30  62]]
0.7758878194490542




### 1.11
Выберите ту модель, на которой метрика recall давала лучший ответ, и рассчитайте для нее precision, используя встроенную функцию.

In [157]:
from sklearn.metrics import precision_score

In [158]:
precision_score(y_test, lr_model_pred)

0.7948717948717948